packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -762,9 +762,16 @@ let dynamicImportedComponentCounter = 0;
 /**
  * create a dynamic component via `<x-foo lwc:dynamic={Ctor}>`
  */
-function dc(
+type LightningElementConfig = {
+    constructor: LightningElementConstructor;
+    props: Record<string, any>;
+};
+function isLightningElementConfig(obj: any): obj is LightningElementConfig {
+    return !!obj.props;
```

Comment:
this seems too be too naive! `class foo extends LightningElement { static props = 1 }` breaks this. We can chat about alternatives, including an explicit mechanism to define the value of a dynamic directive.